core/drand/impl/beaconizer.cpp

```diff
@@ -0,0 +1,212 @@
+    return instance;
+  }
+
+  BeaconizerImpl::BeaconizerImpl(uint64_t filecoin_genesis_time,
+                                 uint64_t filecoin_round_time,
+                                 std::vector<std::string> drand_servers,
+                                 crypto::bls::PublicKey network_public_key,
+                                 size_t max_cache_size)
+      : fil_gen_time_{filecoin_genesis_time},
+        fil_round_time_{filecoin_round_time},
+        peers_{std::move(drand_servers)},
+        network_key_{std::move(network_public_key)},
+        cache_{max_cache_size},
+        bls_{std::make_unique<crypto::bls::BlsProviderImpl>()} {
+    BOOST_ASSERT(not peers_.empty());
+    BOOST_ASSERT(max_cache_size > 0);
+    rotatePeersIndex();
+    dial();
+  }
+
+  outcome::result<BeaconEntry> BeaconizerImpl::entry(uint64_t round) {
+    BeaconEntry entry{.round = round};
+    auto bytes = lookupCache(round);
+    if (bytes) {
+      entry.data = std::move(bytes.value());
+      return entry;
+    }
+
+    auto response = client_->publicRand(round);
+    if (response) {
+      entry.data = std::move(response.value().signature);
+      return entry;
+    }
+
+    OUTCOME_TRY(init());
+    // this place looks weird in lotus too :)
+    return response.error();
+  }
+
+  outcome::result<void> BeaconizerImpl::verifyEntry(
+      const BeaconEntry &current, const BeaconEntry &previous) {
+    if (0 == previous.round) {
+      return outcome::success();
+    }
+    OUTCOME_TRY(is_valid,
+                verifyBeaconData(current.round, current.data, previous.data));
+    if (not is_valid) {
+      return Error::INVALID_BEACON;
+    }
+    cacheEntry(current.round, current.data);
+    return outcome::success();
+  }
+
+  outcome::result<uint64_t> BeaconizerImpl::MaxBeaconRoundForEpoch(
+      ChainEpoch fil_epoch) {
+    if (fil_epoch < 0) {
+      return Error::NEGATIVE_EPOCH;
+    }
+    auto epoch = static_cast<uint64_t>(fil_epoch);
+
+    // sometimes the genesis time for filecoin is zero and this goes negative
+    auto latest_ts =
+        (epoch * fil_round_time_ + fil_gen_time_) - fil_round_time_;
+    auto dround = (latest_ts - drand_gen_time_) / drand_interval_;
+    return dround;
+  }
+
+  // private stuff goes below
+
+  outcome::result<void> BeaconizerImpl::init() {
+    rotatePeersIndex();
+    dial();
+    OUTCOME_TRY(group, client_->group());
+    if (group.dist_key.empty()) {
+      return Error::NO_PUBLIC_KEY;
+    }
+    OUTCOME_TRY(verifyNetworkKey(group.dist_key[0]));
+    drand_interval_ = group.period;
+    drand_gen_time_ = group.genesis_time;
+    return outcome::success();
+  }
+
+  outcome::result<void> BeaconizerImpl::verifyNetworkKey(
+      const Bytes &key) const {
+    auto &key1 = network_key_;
+    if (key.size() != key1.size()) {
+      return crypto::bls::Errors::InvalidPublicKey;
+    }
+    if (not std::equal(key1.begin(), key1.end(), key.begin())) {
+      return Error::NETWORK_KEY_MISMATCH;
+    }
```

Comment:
Maybe do via `==`

Reply:
not for this case because here we have std::array on the one side and std::vector on the other